### ISSUE TITLE
non-result throws

### ISSUE THREAD

**Issue by withoutboats** [withoutboats/fehler#27]
Would like to support:

```rust
#[throws(as Option)]
```

etc

The problem is with the `throw!` macro. If the try trait were stable, we could instead run this:

```rust
return <_ as core::ops::Try>::from_err(..)
```

But that would make this a nightly only crate.

I'm not sure if 

## Design notes

I'd make the syntax work like this:

```rust
#[throws(as $wrap)]
fn function() -> $ret

// translated to:
fn function() -> $wrap<$ret>
```

Whereas:

```rust
#[throws($err as $wrap)]
fn function() -> $ret

// translates to:

fn function() -> $wrap<$ret, $err>
```

This is why I modified the default error type feature to require an explicit `_`.

**Comment by withoutboats** [comment]
We could just put this under a `nightly` flag, but I'm concerned that changing the expansion of `throws!` to involve another type variable will result in inference failures and be a breaking difference between the non-nightly and nightly versions.

**Comment by withoutboats** [comment]
Since `Try` is unstable, we could consider only supporting the stable types somehow. But I don't know how we would do that since that is not just `Result` and `Option` but also `Poll<Result>` and `Poll<Option<Result<>>>`.

We could also consider only supporting `Option` and `Result` and have no integration with the `Try` trait.

**Comment by withoutboats** [comment]
I've got a branch implementing this, but I've encountered what may be a current limitation of rustc: it seems that between the parens in a proc macro attribute cannot be an arbitrary tokenstream, and `#[throws(X as Y)]` does not parse.

**Comment by withoutboats** [comment]
However, I did get throw! to work in a nightly vs stable compatible way I believe, using this:

```rust
macro_rules! throw {
    ($err:expr)   => (return <_ as $crate::__internal::_Try>::from_error((From::from($err))));

    // This is for `Option`, to be able to `throw!()` with no arguments.
    // It would be better if this were `from_error(Default::default())`, but
    // NoneError does not implement `Default` right now.
    ()            => (return <_ as Default>::default());
}

#[doc(hidden)]
pub mod __internal {
    pub trait _Try {
        type Error;
        fn from_error(error: Self::Error) -> Self;
    }

    // Should be a strict and identical subset of the set of instantiations
    // achieved by the nightly module.
    #[cfg(not(feature = "nightly"))]
    mod stable {
        use core::task::Poll;

        impl<T, E> super::_Try for Result<T, E> {
            type Error = E;
            fn from_error(error: Self::Error) -> Self {
                Err(error)
            }
        }

        impl<T, E> super::_Try for Poll<Result<T, E>> {
            type Error = E;

            fn from_error(error: Self::Error) -> Self {
                Poll::Ready(Err(error))
            }
        }

        impl<T, E> super::_Try for Poll<Option<Result<T, E>>> {
            type Error = E;

            fn from_error(error: Self::Error) -> Self {
                Poll::Ready(Some(Err(error)))
            }
        }
    }

    #[cfg(feature = "nightly")]
    mod nightly {
        use core::ops::Try;

        impl<T> super::_Try for T where T: Try {
            type Error = <T as Try>::Error;
            fn from_error(error: Self::Error) -> Self {
                <T as Try>::from_error(error)
            }
        }
    }
}
```

**Comment by withoutboats** [comment]
Closed by #31
